AWS Go support

yunspace opened the pull request:
## What did you implement:

Closes #4507 

[AWS Lambda supports Go now](https://aws.amazon.com/about-aws/whats-new/2018/01/aws-lambda-supports-go/), serverless should support it too.

I will leave the local invoke to another PR. I think I know how to do it.

Note I called the plugin `aws-go` instead of `aws-golang` to be consistent with AWS library name: https://github.com/aws/aws-lambda-go

## How did you implement it:

I leveraged @mthenw's template https://github.com/serverless/serverless-golang which did all the work for me, and then followed @horike37's step by step instructions in the above issue.

There are 2 templates:

1. aws-go - vanilla `go get`
2. aws-go-dep - uses `go dep` 

Each template contains two functions: `hello/main.go` and `world/main.go`

## How can we verify it:

Run `npm test` to make sure all tests pass.

Run the following commands to make sure you can create a go lambda using both templates:

    sls create -t aws-go -p myservice

Or

    sls create -t aws-go-dep -p myservice

Run the following to build and deploy:

    cd aws-go
    ./build.sh
    sls deploy
    sls invoke -f hello
    sls invoke -f world

Run integration test:

    integration-test aws-go 'cd /go/src/app && ./build.sh'
    integration-test aws-go-dep 'cd /go/src/app && ./build.sh'

## Todos:

- [X] Write tests
- [X] Write documentation
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

JosephWoodward commented:
Awesome, looking forward to this 👍

mthenw commented:
LGTM!

HyperBrain commented:
Cool! The template and sample only show exactly one go function in the service. It would be great to have at least 2 in there, so that one can see how to setup a real-world service with Go.

yunspace commented:
@HyperBrain yes I agree. However, I did notice the other vanilla templates within this repo are all quite simple (all hello worlds). I propose the following:

1. an extra example with `go deps` instead of `go get`, similar to `aws-java-maven` and `aws-java-gradle` to go inside this repo.
2. a multi-functions example with API-gateway integration to go inside the more detailed examples repo: https://github.com/serverless/examples 

What does everyone think?

Now that `go` is officially supported, I also have several examples I can bring over from https://github.com/yunspace/serverless-golang

franciscocpg commented:
I agree with @HyperBrain and liked your proposal  @yunspace.

franciscocpg commented on docs/providers/aws/examples/hello-world/go/README.md:
```diff
@@ -0,0 +1,42 @@
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/examples/hello-world/go/)
+<!-- DOCS-SITE-LINK:END -->
+
+# Hello World C# Example
```
Here would be `Hello World Go Example`?

yunspace replied:
thanks, fixed

yunspace commented:
Sorry I totally misunderstood, the idea is to have multiple functions, not multiple examples. Having a look into this now

yunspace commented:
This is ready for review. All unit and integration test are passing for both templates.

yunspace commented:
Hi all, I've taken some inspiration from @mthenw's https://github.com/mthenw/serverless-golang-template and made the following minor changes:

- use `make` instead of duplicated build scripts. This is more elegant and an accepted pattern amongst gophers.
- shorten the binary output from `bin/hello/main` to just `bin/hello`

franciscocpg commented on lib/plugins/create/templates/aws-go/gitignore:
```diff
@@ -0,0 +1,4 @@
+bin
```
IMO we should document this folder as well.
Suggestion: `# golang output binary directory`

franciscocpg commented on lib/plugins/create/templates/aws-go-dep/gitignore:
```diff
@@ -0,0 +1,5 @@
+
+# Serverless directories
+.serverless
+vendor
```
IMO we should document this folder as well.
Something like
```.gitignore
# Serverless directories
.serverless

# golang output binary directory
bin

# golang vendor (dependencies) directory
vendor 
```

yunspace replied:
ah yes sure thing

franciscocpg commented on docs/providers/aws/examples/hello-world/go/README.md:
```diff
@@ -0,0 +1,60 @@
+<!--
```
For `aws-go-dep` example, it only works if we are within a `$GOPATH/src` folder.

I'm just wondering if maybe we should document it somewhere and refer to https://golang.org/doc/code.html#GOPATH?

yunspace replied:
true, these templates and docs are aimed at first time users. It would be better to make them more beginner-friendly. Will update

yunspace commented:
I've made the changes based on @franciscocpg's feedback.

One thing that still confuses me is, why do we have source code with the docs/providers/aws/examples folder? These codes are not accessible from here https://serverless.com/framework/docs/providers/aws/examples/hello-world/ and are identical to the template code. But now require maintenance in two places.

horike37 commented:
Good catch @yunspace :+1: 
I'm not sure the clear reason. However, we don't need to put the example source code there as well.
Actually, we don't see much harm without it.

yunspace commented:
@horike37 thanks for clarifying. I've removed the go code from the docs and left the README there. Stale code won't get updated and will confuse people.

@HyperBrain I've squashed everything into one commit. I really hope the travis doesn't time out on me again...

horike37 commented:
@yunspace 
Awesome! Thank you for the update :+1: 
Just now all the things to land into master has been completed :tada: :100: 

This template will be shipped in the next release!
Hitting the merge button now!
